### PR TITLE
Bookmarking info

### PR DESCRIPTION
**Pull request by gogonzo**
Implement `teal_bookmarable` 

<details>
<summary>Try the app </summary>

```r
options(teal.log_level = "TRACE", teal.show_js_log = TRUE)
# set bookmarking option to a "link"
pkgload::load_all("teal")
pkgload::load_all("teal.modules.general")
# shiny::shinyOptions("bookmarkStore" = "url")

nest_logo <- "https://raw.githubusercontent.com/insightsengineering/hex-stickers/main/PNG/nest.png"

data <- teal_data()
data <- within(data, {
  IRIS <- iris
  MTCARS <- mtcars
})

app <- init(
  data = data,
  filter = teal_slices(
    teal_slice(dataname = "IRIS", varname = "Species", multiple = FALSE)
  ),
  modules = modules(
    modules(
      label = "nested in a tab",
      example_module(),
      teal.modules.general::tm_data_table(),
      teal.modules.general::tm_variable_browser()
    )
  )
)

shinyApp(app$ui, app$server)

```

</details>

image

### CODE REVIEW

File: R/module_bookmark_manager.R [hunk 1]
```diff
@@ -53,149 +29,99 @@
...
-      actionLink(ns("bookmark_add"), NULL, icon = suppressMessages(icon("solid fa-bookmark")), title = "add bookmark"),
-      NULL
-    ),
-    uiOutput(ns("bookmark_list"))
+
+  is_unbookmarkable <- rapply(
+    get_teal_bookmarkable_summary(modules),
+    Negate(isTRUE),
+    how = "unlist"
+  )
```
Comment (chlebowa) [inline R/module_bookmark_manager.R:42]: If the button is created in the server, this part can be removed and the ui function need not take the `modules` argument.

File: R/module_bookmark_manager.R [hunk 2]
```diff
@@ -53,149 +29,99 @@
...
 
     # Set up bookmarking callbacks ----
     # Register bookmark exclusions: all buttons and the `textInput` for bookmark name.
-    setBookmarkExclude(c("bookmark_add", "bookmark_accept", "bookmark_name"))
-    # Add bookmark history to bookmark.
-    session$onBookmark(function(state) {
-      logger::log_trace("bookmark_manager_srv@onBookmark: storing bookmark history")
-      state$values$bookmark_history <- bookmark_history() # isolate this?
-    })
+    setBookmarkExclude(c("do_bookmark"))
```
Comment (chlebowa) [inline R/module_bookmark_manager.R:76]: Modify comments accordingly.

File: R/module_bookmark_manager.R (hunk 1 above)
Comment (gogonzo) [reply R/module_bookmark_manager.R]: done

File: R/module_bookmark_manager.R [hunk 3]
```diff
@@ -53,149 +29,99 @@
 #' @name module_bookmark_manager
 #' @aliases bookmark bookmark_manager bookmark_manager_module
 #'
-
 #' @rdname module_bookmark_manager
 #' @keywords internal
 #'
-bookmark_manager_ui <- function(id) {
+bookmark_module_ui <- function(id, modules) {
```
[the remark this answers is not in the document]
Comment (gogonzo) [reply R/module_bookmark_manager.R]: My bad, thanks

File: R/module_bookmark_manager.R (hunk 2 above)
Comment (gogonzo) [reply R/module_bookmark_manager.R]: Done

File: R/module_bookmark_manager.R [hunk 4]
```diff
@@ -6,196 +6,129 @@
...
-        div(
-          class = "manager_placeholder",
-          "Bookmarks will appear here."
-        )
-      } else {
-        rows
-      }
-    })
-
-    bookmark_history
```
Comment (chlebowa) [inline R/module_bookmark_manager.R:198]: Maybe keep a return value for the sake of testing? If yes, rename properly here and in `wunder_bar_srv` accordingly.

File: inst/css/sidebar.css [hunk 5]
```diff
@@ -1,13 +1,28 @@
...
 
+.badge-count {
+  padding-left: 9px;
+  padding-right: 9px;
+  -webkit-border-radius: 9px;
+  -moz-border-radius: 9px;
+  border-radius: 9px;
+  font-size: 12px;
+  background: #ff0000;
+  color: #fff;
```
Comment (chlebowa) [inline inst/css/sidebar.css:20]: In `teal.slice` we use `var(--bs-gray-200, #e9ecef)` to have a theme color with a hard coded fallback. Could you apply this here as well?

File: inst/css/sidebar.css [hunk 6]
```diff
@@ -1,13 +1,28 @@
...
   background-color: transparent !important;
 }
 
+.badge-count {
+  padding-left: 9px;
+  padding-right: 9px;
+  -webkit-border-radius: 9px;
+  -moz-border-radius: 9px;
+  border-radius: 9px;
+  font-size: 12px;
```
Comment (chlebowa) [inline inst/css/sidebar.css:18]: Consider using `rem` units (emphasis on consider).

File: inst/css/sidebar.css [hunk 7]
```diff
@@ -26,13 +41,16 @@ a.disabled {
...
+
+.manager_table_row *+* {
   flex: 0 0 0px;
   padding: 0em 1.5em;
 }
+
 .manager_placeholder {
   margin-top: 1em;
-}
+}
```
Comment (chlebowa) [inline inst/css/sidebar.css:56]: What is this? :thinking: 
image

**Review by chlebowa — CHANGES_REQUESTED** [review]
Please remember to modify unit tests.

File: R/module_bookmark_manager.R [hunk 8]
```diff
@@ -346,3 +279,21 @@ bookmarks_identical <- function(book1, book2) {
...
+#' @keywords internal
+get_teal_bookmarkable_summary <- function(modules) {
+  checkmate::assert_multi_class(modules, c("teal_modules", "teal_module"))
+  if (inherits(modules, "teal_modules")) {
+    setNames(
+      lapply(modules$children, get_teal_bookmarkable_summary),
+      vapply(modules$children, `[[`, "label", FUN.VALUE = character(1))
+    )
+  } else {
+    isTRUE(attr(modules, "teal_bookmarkable"))
```
Comment (chlebowa) [inline R/module_bookmark_manager.R:297]:
```suggestion
    attr(modules, "teal_bookmarkable", exact = TRUE)
```

File: R/module_bookmark_manager.R [hunk 9]
```diff
@@ -6,196 +6,137 @@
...
-          "Please contact your app developer."
+    is_unbookmarkable <- rapply(
+      modules_bookmarkable(modules),
+      Negate(isTRUE),
+      how = "unlist"
+    )
+    # Render bookmark warnings count
+    output$bookmark_warning <- renderUI({
+      if (!identical(bookmark_option, "server")) {
+        shinyjs::hide("do_bookmark")
```
Comment (chlebowa) [inline R/module_bookmark_manager.R:65]: Hiding the button renders the warning in `modal_content` obsolete, doesn't it?

File: R/modules.R [hunk 10]
```diff
@@ -421,3 +421,21 @@ module_labels <- function(modules) {
...
+#' @keywords internal
+modules_bookmarkable <- function(modules) {
+  checkmate::assert_multi_class(modules, c("teal_modules", "teal_module"))
+  if (inherits(modules, "teal_modules")) {
+    setNames(
+      lapply(modules$children, modules_bookmarkable),
+      vapply(modules$children, `[[`, "label", FUN.VALUE = character(1))
+    )
+  } else {
+    isTRUE(attr(modules, "teal_bookmarkable", exact = TRUE))
```
Comment (chlebowa) [inline R/modules.R:439]: This drops `NULL`s which would be handled differently to `FALSE`s later on.

File: R/module_bookmark_manager.R (hunk 9 above)
Comment (gogonzo) [reply R/module_bookmark_manager.R:65]: This thing is not in a modal - this is a warning badge. I made small changes anyway

File: R/modules.R (hunk 10 above)
Comment (gogonzo) [reply R/modules.R]: good point

File: R/module_bookmark_manager.R (hunk 9 above)
Comment (chlebowa) [reply R/module_bookmark_manager.R:65]: My mistake.

**Review by chlebowa — APPROVED** [review]

**Pull request merged**